libraries/chain/include/eosio/chain/finality/finalizer_policy.hpp

```diff
@@ -2,13 +2,38 @@
 
 #include <eosio/chain/types.hpp>
 #include <eosio/chain/finality/finalizer_authority.hpp>
+#include <fc/container/ordered_diff.hpp>
 
 namespace eosio::chain {
 
+   static_assert(std::numeric_limits<uint16_t>::max() <= config::max_finalizers);
+   using finalizers_differ = fc::ordered_diff<finalizer_authority, uint16_t>;
+   using finalizers_diff_t = finalizers_differ::diff_result;
+
+   struct finalizer_policy_diff {
+      uint32_t                         generation = 0; ///< sequentially incrementing version number
+      uint64_t                         threshold = 0;  ///< vote weight threshold to finalize blocks
+      finalizers_diff_t                finalizers_diff;
+   };
+
    struct finalizer_policy {
       uint32_t                         generation = 0; ///< sequentially incrementing version number
       uint64_t                         threshold = 0;  ///< vote weight threshold to finalize blocks
-      std::vector<finalizer_authority> finalizers; ///< Instant Finality voter set
+      std::vector<finalizer_authority> finalizers;     ///< Instant Finality voter set
+
+      finalizer_policy_diff create_diff(const finalizer_policy& target) const {
+         return {.generation = target.generation,
+                 .threshold = target.threshold,
+                 .finalizers_diff = finalizers_differ::diff(finalizers, target.finalizers)};
+      }
+
+      template <typename X>
+      requires std::same_as<std::decay_t<X>, finalizer_policy_diff>
+      void apply_diff(X&& diff) {
+         generation = diff.generation;
+         threshold = diff.threshold;
+         finalizers_differ::apply_diff(finalizers, std::move(diff.finalizers_diff));
```

Comment:
That's wrong, it should be `std::forward<X>(diff).finalizers_diff`.

`std::move` will move unconditionally from the in-flight policy diffs stored in  `block_header_state:: finalizer_policies` in `calculate_finalizer_policy_diff`